File: frontend/src/utils/moleculeComparison.js
/**
 * Funções para comparar moléculas considerando estrutura completa
 * (partículas + ligações)
 */

/**
 * Compara duas moléculas verificando se são idênticas estruturalmente
 * Considera: tipos de partículas, polaridades E estrutura de ligações
 */
export function areMoleculesIdentical(mol1, mol2) {
  if (!mol1 || !mol2) return false
  
  // Verificar quantidade de partículas
  if (mol1.particles.length !== mol2.particles.length) return false
  
  // Verificar quantidade de ligações
  if (mol1.bonds.length !== mol2.bonds.length) return false
  
  // Criar "impressão digital" molecular (fingerprint)
  const fingerprint1 = createMolecularFingerprint(mol1)
  const fingerprint2 = createMolecularFingerprint(mol2)
  
  return fingerprint1 === fingerprint2
}

/**
 * Cria uma "impressão digital" única da molécula
 * baseada na estrutura completa (partículas + ligações)
 */
function createMolecularFingerprint(molecule) {
  // 1. Criar representação das partículas (ordenada)
  const particlesSignature = molecule.particles
    .map(p => `${p.type}${p.polarity}`)
    .sort()
    .join(',')
  
  // 2. Criar representação das ligações (normalizada)
  // Cada ligação é representada pelos tipos/polaridades conectados
  const bondsSignature = molecule.bonds
    .map(bond => {
      const particleFrom = molecule.particles.find(p => p.id === bond.from)
      const particleTo = molecule.particles.find(p => p.id === bond.to)
      
      if (!particleFrom || !particleTo) return ''
      
      const from = `${particleFrom.type}${particleFrom.polarity}`
      const to = `${particleTo.type}${particleTo.polarity}`
      const mult = `x${bond.multiplicity}`
      
      // Ordenar para que A-B seja igual a B-A
      const [first, second] = [from, to].sort()
      return `${first}-${second}${mult}`
    })
    .filter(b => b) // Remover vazios
    .sort()
    .join('|')
  
  // 3. Combinar tudo em uma única string
  return `${particlesSignature}::${bondsSignature}`
}

/**
 * Verifica se uma molécula já existe numa lista
 */
export function moleculeExistsInList(molecule, moleculeList) {
  return moleculeList.some(mol => areMoleculesIdentical(molecule, mol))
}

File: backend/core/molecule_comparison.py
"""
Funções para comparar moléculas considerando estrutura completa
(partículas + ligações)
"""

def are_molecules_identical(mol1, mol2):
    """
    Compara duas moléculas verificando se são idênticas estruturalmente
    Considera: tipos de partículas, polaridades E estrutura de ligações
    """
    if not mol1 or not mol2:
        return False
    
    # Verificar se são dicionários (moléculas)
    if not isinstance(mol1, dict) or not isinstance(mol2, dict):
        return False
    
    # Verificar quantidade de partículas
    if len(mol1.get('particles', [])) != len(mol2.get('particles', [])):
        return False
    
    # Verificar quantidade de ligações
    if len(mol1.get('bonds', [])) != len(mol2.get('bonds', [])):
        return False
    
    # Criar "impressão digital" molecular (fingerprint)
    fingerprint1 = create_molecular_fingerprint(mol1)
    fingerprint2 = create_molecular_fingerprint(mol2)
    
    return fingerprint1 == fingerprint2


def create_molecular_fingerprint(molecule):
    """
    Cria uma "impressão digital" única da molécula
    baseada na estrutura completa (partículas + ligações)
    """
    particles = molecule.get('particles', [])
    bonds = molecule.get('bonds', [])
    
    # 1. Criar representação das partículas (ordenada)
    particles_signature = ','.join(sorted([
        f"{p['type']}{p['polarity']}" for p in particles
    ]))
    
    # 2. Criar representação das ligações (normalizada)
    bonds_signature_parts = []
    for bond in bonds:
        particle_from = next((p for p in particles if p['id'] == bond['from']), None)
        particle_to = next((p for p in particles if p['id'] == bond['to']), None)
        
        if not particle_from or not particle_to:
            continue
        
        from_str = f"{particle_from['type']}{particle_from['polarity']}"
        to_str = f"{particle_to['type']}{particle_to['polarity']}"
        mult = f"x{bond.get('multiplicity', 1)}"
        
        # Ordenar para que A-B seja igual a B-A
        first, second = sorted([from_str, to_str])
        bonds_signature_parts.append(f"{first}-{second}{mult}")
    
    bonds_signature = '|'.join(sorted(bonds_signature_parts))
    
    # 3. Combinar tudo em uma única string
    return f"{particles_signature}::{bonds_signature}"


